src/python/parla/cython/device.pyx

```diff
@@ -0,0 +1,126 @@
+from parla.cython import core
+
+cdef class CyDeviceManager:
+    """
+    This class manages devices on the current system.
+    For convenience, it registers devices specified by users to both
+    pure Python and Cython side; So, it replicates processes.
+    """
+    cdef DeviceManager* cpp_device_manager_
+
+    def __cinit__(self):
+        self.cpp_device_manager_ = new DeviceManager()
+
+    def __init__(self):
+        pass
+
+    def __dealloc__(self):
+        del self.cpp_device_manager_
+
+    cpdef register_device(self, CyDevice cy_device):
+        cdef Device* cpp_device = cy_device.get_cpp_device()
+        self.cpp_device_manager_.RegisterDevice(cpp_device)
+
+    cpdef print_registered_devices(self):
+        self.cpp_device_manager_.PrintRegisteredDevices()
+
+
+
+class PyDevice:
+
+    def __init__(self, dev_type_name, dev_id: int):
+        self._device_name = dev_type_name + ":" + str(dev_id)
+        print(f"Device: {self._device_name} is registered.", flush=True)
+
+    def __enter__(self):
+        print(f"Entered device, {self.get_name()}, context", flush=True)
+
+    def __exit__(self):
+        print(f"Exited device, {self.get_name()}, context", flush=True)
+
+    def get_name(self):
+        return self._device_name
+
+    def get_cy_device(self):
+        return self._cy_device
+
+    def __repr__(self):
+        return self._device_name
+
+
+"""
+Device instances in Python manage resource status.
+TODO(hc): the device configuration will be packed in a data class soon.
+"""
+
+class PyCUDADevice(PyDevice):
+
+    def __init__(self, dev_id: int, mem_sz: long, num_vucs: long):
+        super().__init__("CUDA", dev_id)
+        self._cy_device = CyCUDADevice(dev_id, mem_sz, num_vucs, self)
+
+
+class PyCPUDevice(PyDevice):
+
+    def __init__(self, dev_id: int, mem_sz: long, num_vucs: long):
+        super().__init__("CPU", dev_id)
+        self._cy_device = CyCPUDevice(dev_id, mem_sz, num_vucs, self)
+
+
+cdef class CyDevice:
+    cdef Device* _cpp_device
+
+    cdef Device* get_cpp_device(self):
+        return self._cpp_device
+
+
+cdef class CyCUDADevice(CyDevice):
+
+    def __cinit__(self, int dev_id, long mem_sz, long num_vucs, py_device):
+        self._cpp_device = new CUDADevice(dev_id, mem_sz, num_vucs, <void *> py_device)
```

Comment:
This is good for testing but side comment: Should we keep the resource types more general instead of hardcoding memory size/vcus and getters for those? (Keep the getters/setters through the arbitrary string map like in current Parla)

Reply:
Good point. Yes that's my next Todo :-)